fe/parameterize_batch.py

```diff
@@ -0,0 +1,385 @@
+
+if __name__ == "__main__":
+
+    parser = argparse.ArgumentParser(description='Quick Test')
+    parser.add_argument('--frames_dir', type=str, required=True)
+    parser.add_argument('--precision', type=str, required=True)    
+    parser.add_argument('--complex_pdb', type=str, required=True)
+
+    """The ligands in ligand_sdf_file should be listed in the same order as the true_dG_file"""
+    parser.add_argument('--ligand_sdf_file', type=str, nargs='*')
+    parser.add_argument('--num_gpus', type=int, required=True)
+    parser.add_argument('--jobs_per_gpu', type=int, required=True)
+    parser.add_argument('--num_conformers', type=int, required=True)
+
+    """The dG values are in kJ/mol and positive sign for both insertion and deletion (1kcal/mol=4.184kJ/mol)"""
+    parser.add_argument('--true_dG_file', type=str, nargs='*')
+
+    args = parser.parse_args()
+
+    assert os.path.isdir(args.frames_dir)
+
+    if args.precision == 'single':
+        precision = np.float32
+    elif args.precision == 'double':
+        precision = np.float64
+    else:
+        raise Exception("precision must be either single or double")
+
+    num_gpus = args.num_gpus
+    num_workers = args.num_gpus*args.jobs_per_gpu
+
+    print('Creating multiprocessing pool with',args.num_gpus, 'gpus and', args.jobs_per_gpu, 'jobs per gpu')
+    pool = multiprocessing.Pool(num_workers)
+
+    start = 1e3
+    end = 1.0
+    NT = 500
+    base = np.exp(np.log(end/start)/NT)
+    exps = np.arange(NT)
+    part_one = np.power(base, exps)*start
+    part_two = np.linspace(1.0, 0.3, 1000)
+    part_three = np.linspace(0.3, 0.0, 4000)
+
+    forward_schedule = np.concatenate([part_one, part_two, part_three])
+    backward_schedule = forward_schedule[::-1]
+    lambda_schedule = np.concatenate([forward_schedule, backward_schedule])
+
+    T = lambda_schedule.shape[0]
+    assert T % 2 == 0
+    dt = 0.001
+    step_sizes = np.ones(T)*dt
+    assert T % 2 == 0
+    cas = np.ones(T)*0.93
+    epoch = 0
+    lr = 5e-4
+
+    ligand_list = []
+    for ligand_sdf_file, ligand_dG_file in zip(args.ligand_sdf_file, args.true_dG_file):
+        ligand_sdf_file_readlines = open(ligand_sdf_file, 'r').readlines()
+        ligand_dG_file_readlines = open(ligand_dG_file, 'r').readlines()
+
+        for file_name, deltaG in zip(ligand_sdf_file_readlines, ligand_dG_file_readlines):
+            
+            for rd_mol in Chem.SDMolSupplier(file_name.rstrip('\n'), removeHs=False):
+                break
+
+            true_dG_value = float(deltaG.split()[0])
+            ligand_list.append((rd_mol, true_dG_value))
+
+    guest_mol = ligand_list[0][0]
+        
+    off = smirnoff.ForceField("test_forcefields/smirnoff99Frosst.offxml")
+    _, (guest_params, _), _ = forcefield.parameterize(guest_mol, off, dimension=4)
+    
+    opt_init, opt_update, get_params = optimizers.adam(lr)
+    opt_state = opt_init(guest_params)
+
+    np.savetxt("original_off_params.txt", guest_params)
```

Comment:
you probably want to call this again after the end of every epoch.

Reply:
for your own sake, probably best to also save the corresponding guest_param_groups